### DISCUSSION TITLE
left right wave extension

### DISCUSSION THREAD

**Discussion by spamatica** [muse-sequencer/muse#1215] [Team Posts]
Not really sure how to handle this chat, as kybos said it is very hard to find prior discussions, I haven't find a search either... made a new post now anyway, not sure it is the right way but there is no clear right...

Anyway, I did some more experimenting and found out more how it works on my system (and a workaround). 
Not sure I understood your last explanation Tim, that it only one way should have a hidden event, but in any case I could not extend it at the cut where there should have been more wave file.

I attached a short video that shows how it appears on my system so we can compare notes.

No sound in the file so I'll write down what I did
1. import a wave file
2. cut it, 
3. it gets a gray hidden event marker on the left part but not on the right.
4. the left one can be extended but the right one cannot be extended 
5. undo it back to the full wave
6. resize it at the beginning (this is the work around i mentioned)
7. cut it again
8. now both parts have a gray hidden event marker and can both be extended.

There was also a moment where I thought a second cut didn't allow it to be extended to the left but I couldn't reproduce it. Either I got confused or there is a bit of randomness to this.

https://user-images.githubusercontent.com/4476983/116092568-3346ec00-a6a6-11eb-8a3b-191d54e365a4.mp4

**Comment by terminator356** [comment]
Unable to watch. No H264 decoders here.

I know what you are looking for.
A way to extend the wave event so that its 'spos' offset member is adjusted.
I know you had the code in there before I changed it.
But at this moment there absolutely is no way to adjust a wave event's 'spos' offset.
This has been missing since day one.

However I will be adding some more code to adjust it in the wave editor.
The Ctrl key is free and available while adjusting the wave event's borders, so I will use that
 to allow extending the wave event's 'spos' value. (Very similar to the use of Ctrl key in the arranger
 to allow 'dragging' the events with the borders, except here it will adjust the wave event's 'spos' value.)
I will also be adding small 'hidden' markers just for the wave events, not only in the wave editor
 but also in the arranger, to alert users that the wave event has an offset (spos member).

I will try to add the ability to adjust this 'spos' value in the arranger as well if I can find a key to use
 but as I mentioned before, arranger is not the best place to be doing that.
I know you are trying to make it easy to recover hidden parts of the wave from within the arranger,
 so I will do my best to somehow accommodate that, but don't expect miracles...
When adjusting borders in the arranger, since there aren't any real options left over
 for key combinations because Shift and Ctrl are used, my thoughts are leaning towards
 having a 'mode' button or box on the toolbar.

Hang in there, I will do my best...

**Comment by terminator356** [comment]
Step 1 Done:
In the wave editor, hold the Ctrl key while dragging event borders to drag the contents with the border.
In other words, if dragging the left border while **not** holding the Ctrl key, the wave event's 'spos' offset value
 can be adjusted. In other words it exposes more of the wave. Same deal with right border.
Similar to how the arranger parts can 'drag' the events along with the borders by holding the Ctrl key.

I have limited the wave event 'spos' offset value to spos >= 0.
Because I'm not sure yet that any part of the app ever set it at a negative value ie. a gap between
 left event border and start of wave.
So you'll notice when dragging the borders it can only expose so much of the wave at left.

I know it's more clumsy than allowing adjustment in the arranger, but hey it's a start.
This is the first time that the 'spos' value can actually be adjusted in the wave editor !

Step 2: Draw small 'hidden' markers for the wave events. Thinking top corners.
Must re-work those part 'hidden' event markers. Too obscured by the waves, same colour as RMS wave.
Thinking bottom corners. Thinking alternatively about rounded part corners, as you attempted once Robert
 but I stopped you. Perhaps normally rounded, but flat when hidden events?

It's late, gotta go...

**Comment by terminator356** [comment]
Also, I am aware that we have some restrictions in there that 'forbid' moving an event
 inside a part that has 'hidden' markers if that movement would require expanding the part.
In the case of the wave editor this seems a bit excessive.
It can seem frustrating that in the wave editor a wave event refuses to move just because
 the part's 'hidden events' markers are showing.
It might be that this was a cause of frustration for you.

**Comment by terminator356** [comment]
Finally, a slight caveat with that finished Step 1 above:

I discovered that my dynamic 'live' (non-destructive) samplerate converters seem to handle
 the 'spos' offset slightly wrong.
Thus if you are viewing a 48KHz wave with a system rate of 44.1KHz and you've agreed to slap
 one of my 'live' converters on the wave when importing it, then manipulating the wave event's borders
 as described above produces a slight forward offset each time.
If the wave and system samplerates are the same, then there's no problem, adjusting the borders
 produces smooth changes without slight offset drift.

**Comment by spamatica** [comment]
I'm all For you making cool improvements :) but I still think we are not entirely on the same page. 

> But at this moment there absolutely is no way to adjust a wave event's 'spos' offset.
This has been missing since day one.

Yet somehow it works if I apply the workaround, is it the events being moved then?
Can I share the video any other way? Youtube transcodes to many formats, that must work? I'll upload it in a bit.

> I will try to add the ability to adjust this 'spos' value in the arranger as well if I can find a key to use
but as I mentioned before, arranger is not the best place to be doing that.

I don't wish to start a shouting match, it may be that technically this is not perfect. But from a workflow perspective I can't see how this can be done anywhere but the arranger. When there is an opportunity I will try to make a video describing how I work with this. :)

...I see now that if I continue to cut the part, resizing the new edges does not always allow to resize the entire wave, not on the right side either. 
To me this looks like the functionality is pretty much there, but there are some special cases that don't work right.

**Comment by spamatica** [comment]
Here is the same video uploaded to youtube
https://youtu.be/MLYBhRXjMe8

**Comment by terminator356** [comment]
I see what it's doing. It's unintended behaviour. It's supposed to behave exactly like the first half of the video.
Instead it got the right hand part position correct, but got its wave event position and offset wrong.
But it behaved in a way that was what you wanted.

You are seeing behaviour that I was going to mention, and I was sure you had also thought:
"Now that we have part border resizing, what if we eliminated the **wave event offset value** (spos member)
 and simply relied on the borders to make a 'window' on the entire fully expanded wave event?".
It's a great idea - but only if just one wave event were allowed per part. For now we need that offset value.

So I'm going to try something...
